generate/generate.go

```diff
@@ -440,34 +484,110 @@ func writeGeneratedCode(target string, code []byte) error {
 }
 
 func checkFileCanBeOverwritten(path string) error {
+	_, err := loadGeneratedCode(path)
+	return err
+}
+
+func loadGeneratedCode(path string) ([]byte, error) {
 	logger := log.With().
-		Str("action", "checkFileCanBeOverwritten()").
+		Str("action", "loadGeneratedCode()").
 		Str("path", path).
 		Logger()
 
-	logger.Trace().
-		Msg("Get file information.")
+	logger.Trace().Msg("Get file information.")
+
 	_, err := os.Stat(path)
 	if err != nil {
 		if errors.Is(err, fs.ErrNotExist) {
-			return nil
+			return nil, nil
 		}
-		return fmt.Errorf("unsafe to overwrite file, can't stat %q", path)
+		return nil, fmt.Errorf("loading code: can't stat %q: %v", path, err)
 	}
 
-	logger.Trace().
-		Msg("Read file.")
+	logger.Trace().Msg("Read file.")
+
 	data, err := os.ReadFile(path)
 	if err != nil {
-		return fmt.Errorf("unsafe to overwrite file, can't read %q", path)
+		return nil, fmt.Errorf("loading code, can't read %q: %v", path, err)
 	}
 
-	logger.Trace().
-		Msg("Check if code has terramate header.")
-	code := string(data)
-	if !strings.HasPrefix(code, codeHeader) {
-		return fmt.Errorf("%w: at %q", ErrManualCodeExists, path)
+	logger.Trace().Msg("Check if code has terramate header.")
+
+	if !strings.HasPrefix(string(data), codeHeader) {
+		return nil, fmt.Errorf("%w: at %q", ErrManualCodeExists, path)
+	}
+
+	return data, nil
+}
+
+func checkProjectRoot(root string) error {
+	if !filepath.IsAbs(root) {
+		return fmt.Errorf("project's root %q must be an absolute path", root)
+	}
+
+	info, err := os.Lstat(root)
+	if err != nil {
+		return fmt.Errorf("checking project's root directory %q: %v", root, err)
+	}
+
+	if !info.IsDir() {
+		return fmt.Errorf("project's root %q is not a directory", root)
 	}
 
 	return nil
 }
+
+type stackIterator func(
+	stackpath string,
+	metadata terramate.StackMetadata,
+	globals *terramate.Globals,
+) error
+
+func iterateStacks(root string, iter stackIterator) []error {
+	logger := log.With().
+		Str("action", "generate.iterateStacks()").
+		Str("path", root).
+		Logger()
+
+	logger.Trace().Msg("Check project root.")
+
+	if err := checkProjectRoot(root); err != nil {
+		return []error{err}
+	}
+
+	logger.Trace().Msg("Load stacks metadata.")
+
+	metadata, err := terramate.LoadMetadata(root)
+	if err != nil {
+		return []error{err}
+	}
+
+	var errs []error
+
+	for _, stackMetadata := range metadata.Stacks {
+		stackpath := project.AbsPath(root, stackMetadata.Path)
+
+		logger := logger.With().
+			Str("stack", stackpath).
+			Logger()
+
+		logger.Trace().Msg("Load stack globals.")
+
+		globals, err := terramate.LoadStackGlobals(root, stackMetadata)
+		if err != nil {
+			errs = append(errs, fmt.Errorf(
+				"stack %q: %w: %v",
+				stackpath,
+				ErrLoadingGlobals,
+				err))
+			continue
+		}
+
+		logger.Trace().Msg("Calling iterator.")
+		if err := iter(stackpath, stackMetadata, globals); err != nil {
```

Comment:
funny haha

Reply:
I think I was remembering some iter stuff from C++ :laughing: (but it is not what is happening here, so naming is off indeed..but I was thinking about it =P)